pkg/repository/prisma/buffered.go

```diff
@@ -0,0 +1,290 @@
+	internalArr        []*inputWrapper[T, U]
+	sizeOfData         int // size of data in buffer
+	maxDataSizeInQueue int // max number of bytes to hold in buffer before we flush
+
+	l      *zerolog.Logger
+	lock   sync.Mutex
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+type inputWrapper[T any, U any] struct {
+	item     T
+	doneChan chan<- *flushResponse[U]
+}
+
+type IngestBufOpts[T any, U any] struct {
+	MaxCapacity        int                                               `validate:"required,gt=0"`
+	FlushPeriod        time.Duration                                     `validate:"required,gt=0"`
+	MaxDataSizeInQueue int                                               `validate:"required,gt=0"`
+	OutputFunc         func(ctx context.Context, items []T) ([]U, error) `validate:"required"`
+	SizeFunc           func(T) int                                       `validate:"required"`
+	L                  *zerolog.Logger                                   `validate:"required"`
+}
+
+// NewIngestBuffer creates a new buffer for any type T
+func NewIngestBuffer[T any, U any](opts IngestBufOpts[T, U]) *IngestBuf[T, U] {
+
+	inputChannelSize := opts.MaxCapacity
+	if inputChannelSize < 100 {
+		inputChannelSize = 100
+	}
+
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
+
+	return &IngestBuf[T, U]{
+		state:              initialized,
+		maxCapacity:        opts.MaxCapacity,
+		flushPeriod:        opts.FlushPeriod,
+		inputChan:          make(chan *inputWrapper[T, U], inputChannelSize),
+		lastFlush:          time.Now(),
+		internalArr:        make([]*inputWrapper[T, U], 0),
+		sizeOfData:         0,
+		maxDataSizeInQueue: opts.MaxDataSizeInQueue,
+		outputFunc:         opts.OutputFunc,
+		sizeFunc:           opts.SizeFunc,
+		l:                  opts.L,
+		ctx:                ctx,
+		cancel:             cancel,
+	}
+}
+
+func (b *IngestBuf[T, U]) safeAppendInternalArray(e *inputWrapper[T, U]) {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	b.internalArr = append(b.internalArr, e)
+}
+
+func (b *IngestBuf[T, U]) safeFetchSizeOfData() int {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	return b.sizeOfData
+}
+
+func (b *IngestBuf[T, U]) safeIncSizeOfData(size int) {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	b.sizeOfData += size
+
+}
+
+func (b *IngestBuf[T, U]) safeDecSizeOfData(size int) {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	b.sizeOfData -= size
+
+}
+
+func (b *IngestBuf[T, U]) safeSetLastFlush(t time.Time) {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	b.lastFlush = t
+
+}
+
+func (b *IngestBuf[T, U]) safeFetchLastFlush() time.Time {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+	return b.lastFlush
+}
+
+func (b *IngestBuf[T, U]) buffWorker() {
+	for {
+		select {
+		case <-b.ctx.Done():
+			return
+		case e := <-b.inputChan:
+			b.safeAppendInternalArray(e)
+			b.safeIncSizeOfData(b.calcSizeOfData([]T{e.item}))
+
+			if b.safeCheckSizeOfBuffer() >= b.maxCapacity {
+				go b.flush(b.sliceInternalArray())
```

Comment:
I think it would be cleaner to create a goroutine only after all concurrent writes w/ the mutex have been performed. With the way this is written, I think we may call `go b.flush` more than once because it calls `b.safeSetLastFlush` and `b.safeDecSizeOfData`. So instead of calling `go b.flush`, I think we want to refactor the `flush` method to something like:

```go
func (b *IngestBuf[T, U]) flush(items []*inputWrapper[T, U]) {
        b.safeSetLastFlush(time.Now())

	numItems := len(items)
	
	var doneChans []chan<- *flushResponse[U]
	opts := make([]T, numItems)

	for i := 0; i < numItems; i++ {
		opts[i] = items[i].item
		doneChans = append(doneChans, items[i].doneChan)
	}

	b.safeDecSizeOfData(b.calcSizeOfData(opts))

        // SPAWN GOROUTINE HERE
        go func() {
	ctx := context.Background()
	result, err := b.outputFunc(ctx, opts)

	if err != nil {
		for _, doneChan := range doneChans {
			select {
			case doneChan <- &flushResponse[U]{err: err}:
			default:
				b.l.Error().Msgf("could not send error to done chan: %v", err)

			}
		}
		return
	}

	for i, d := range doneChans {
		select {
		case d <- &flushResponse[U]{result: result[i], err: nil}:
		default:
			b.l.Error().Msg("could not send done to done chan")
		}
	}

	b.l.Debug().Msgf("Flushed %d items", numItems)
        }()
}
```

And then just call `b.flush` so that we can set the correct size and last flush time synchronously (which should be extremely fast anyway)

Reply:
sounds good, I'll simplify the logic in the calling for select too